src/bcf/mod.rs

```diff
@@ -132,6 +151,149 @@ impl Drop for Reader {
+            },
+            _ => {
+                Err(IndexedReaderPathError::InvalidPath)
+            }
+        }
+    }
+
+    pub fn from_url(url: &Url) -> Result<Self, IndexedReaderError> {
+        Self::new(&ffi::CString::new(url.as_str()).unwrap())
+    }
+
+    /// Create a new `IndexedReader`.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path. Use "-" for stdin.
+    fn new(path: &ffi::CStr) -> Result<Self, IndexedReaderError> {
+        // Create reader and require existence of index file.
+        let ser_reader = unsafe { htslib::bcf_sr_init() };
+        unsafe { htslib::bcf_sr_set_opt(ser_reader, 0); } // 0: BCF_SR_REQUIRE_IDX
+        // Attach a file with the path from the arguments.
+        if unsafe { htslib::bcf_sr_add_reader(ser_reader, path.as_ptr()) } >= 0 {
+            let header = Rc::new(HeaderView::new(unsafe {
+                htslib::bcf_hdr_dup((*(*ser_reader).readers.offset(0)).header) }));
+            Ok(IndexedReader {
+                inner: ser_reader, header: header, current_region: None
+            })
+        } else {
+            Err(IndexedReaderError::InvalidPath)
+        }
+    }
+
+    pub fn fetch(&mut self, rid: u32, beg: u32, end: u32) -> Result<(), FetchError> {
+        let contig = self.header.rid2name(rid);
+        let contig = ffi::CString::new(contig).unwrap();
+        let contig = contig.as_ptr();
+        if unsafe { htslib::bcf_sr_seek(self.inner, contig, beg as i32) } != 0 {
+            Err(FetchError::Some)
+        } else {
+            self.current_region = Some((rid, beg, end));
+            Ok(())
+        }
+    }
+}
+
+
+impl Read for IndexedReader {
+    fn read(&mut self, record: &mut record::Record) -> Result<(), ReadError> {
+        match unsafe { htslib::bcf_sr_next_line(self.inner) } {
+            0 => {
+                if unsafe { (*self.inner).errnum } != 0 {
+                    Err(ReadError::SyncedBcfReaderError)
+                } else {
+                    Err(ReadError::NoMoreRecord)
+                }
+            }
+            i => {
+                assert!(i > 0, "Must not be negative");
+                // Note that the sync BCF reader has a different interface than the others
+                // as it keeps its own buffer already for each record.  An alternative here
+                // would be to replace the `inner` value by an enum that can be a pointer
+                // into a synced reader or an owning popinter to an allocated record.
+                unsafe {
+                    htslib::bcf_copy(
+                        record.inner,
+                        *(*(*self.inner).readers.offset(0)).buffer.offset(0));
+                }
+
+                record.set_header(self.header.clone());
+
+                match self.current_region {
+                    Some((rid, _start, end)) => {
+                        if record.rid().is_some() && rid == record.rid().unwrap() &&
```

Comment:
One last thing. Why is this necessary? I thought the fetch already fetches the correct region?

Reply:
It only jumps to the start of the region...

Reply:
ok!